shimmingtoolbox/shim/sequencer.py

```diff
@@ -42,838 +38,1063 @@
+            dilation_kernel (str): kernel used to dilate the mask. Allowed shapes are: 'sphere', 'cross', 'line'
+                                    'cube'. See :func:`shimmingtoolbox.masking.mask_utils.dilate_binary_mask` for more
+                                    details.
+            dilation_size (int): Length of a side of the 3d kernel to dilate the mask. Must be odd. For example,
+                                         a kernel of size 3 will dilate the mask by 1 pixel.
+            path_output (str): Path to the directory to output figures. Set logging level to debug to output debug
+                           artefacts.
+            shimwise_bounds (list) : Bounds of the currents for the optimization
+
+        Returns:
+                i (integer) : index of the slice used for the optimization
+                coef (list) : List of the coef to applied for the B0 shimming to this slice
+        """
+        
+        nii_unshimmed = nib.Nifti1Image(optimizer.unshimmed, optimizer.unshimmed_affine)
+
+        # Create mask in the fieldmap coordinate system from the anat roi mask and slice anat mask
+        sliced_mask_resampled = resample_mask(nii_mask_anat, nii_unshimmed, slices_anat[i],
+                                              dilation_kernel=dilation_kernel,
+                                              dilation_size=dilation_size,
+                                              path_output=path_output).get_fdata()
+        # If new bounds are included, change them for each shim
+        if shimwise_bounds is not None:
+            optimizer.set_merged_bounds(shimwise_bounds[i])
+        if np.all(sliced_mask_resampled == 0):
+            return i, np.zeros(optimizer.merged_coils.shape[-1])
+
+        # Optimize using the mask
+        coef = optimizer.optimize(sliced_mask_resampled)
+        return i, coef
+
+    def _plot_currents(self, static, riro=None):
+        """Plot evolution of currents through shims
+        Args:
+            static (numpy.ndarray) : Array with the static coefficients
+            riro (numpy.ndarray) : Array with the riro coefficients
+        """
+        fig = Figure(figsize=(10, 10))
+        ax = fig.add_subplot(111)
+        n_channels = static.shape[1]
+        for i_channel in range(n_channels):
+            ax.plot(static[:, i_channel], label=f"Static channel{i_channel} currents through shim groups")
+        if riro is not None:
+            for i_channel in range(n_channels):
+                ax.plot(riro[:, i_channel], label=f"Riro channel{i_channel} currents through shim groups")
+        ax.set_xlabel('Shim group')
+        ax.set_ylabel('Coefficients (Physical CS [RAS])')
+        ax.legend()
+        ax.set_title("Currents through shims")
+        fname_figure = os.path.join(self.path_output, 'fig_currents.png')
+        fig.savefig(fname_figure)
+        logger.debug(f"Saved figure: {fname_figure}")
+
+
+class ShimSequencer(Sequencer):
```

Comment:
Document the attributes of the class (every time you declare self.xxxxx and that it is not a method)

Reply:
I documented the attributes, it should be good now !